lib/internal/modules/esm/hooks.js

```diff
@@ -0,0 +1,652 @@
+          },
+        );
+      }
+    }
+  }
+
+  /**
+   * Initialize `globalPreload` hooks.
+   */
+  #preload() {
+    for (let i = this.#hooks.globalPreload.length - 1; i >= 0; i--) {
+      const { MessageChannel } = require('internal/worker/io');
+      const channel = new MessageChannel();
+      const {
+        port1: insidePreload,
+        port2: insideLoader,
+      } = channel;
+
+      insidePreload.unref();
+      insideLoader.unref();
+
+      const {
+        fn: preload,
+        url: specifier,
+      } = this.#hooks.globalPreload[i];
+
+      const preloaded = preload({
+        port: insideLoader,
+      });
+
+      if (preloaded == null) { return; }
+
+      const hookErrIdentifier = `${specifier} globalPreload`;
+
+      if (typeof preloaded !== 'string') { // [2]
+        throw new ERR_INVALID_RETURN_VALUE(
+          'a string',
+          hookErrIdentifier,
+          preload,
+        );
+      }
+      const { compileFunction } = require('vm');
+      const preloadInit = compileFunction(
+        preloaded,
+        ['getBuiltin', 'port', 'setImportMetaCallback'],
+        {
+          filename: '<preload>',
+        }
+      );
+      const { BuiltinModule } = require('internal/bootstrap/loaders');
+      // We only allow replacing the importMetaInitializer during preload;
+      // after preload is finished, we disable the ability to replace it.
+      //
+      // This exposes accidentally setting the initializer too late by throwing an error.
+      let finished = false;
+      let replacedImportMetaInitializer = false;
+      let next = this.#importMetaInitializer;
+      try {
+        // Calls the compiled preload source text gotten from the hook
+        // Since the parameters are named we use positional parameters
+        // see compileFunction above to cross reference the names
+        FunctionPrototypeCall(
+          preloadInit,
+          globalThis,
+          // Param getBuiltin
+          (builtinName) => {
+            if (BuiltinModule.canBeRequiredByUsers(builtinName) &&
+                BuiltinModule.canBeRequiredWithoutScheme(builtinName)) {
+              return require(builtinName);
+            }
+            throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
+          },
+          // Param port
+          insidePreload,
+          // Param setImportMetaCallback
+          (fn) => {
+            if (finished || typeof fn !== 'function') {
+              throw new ERR_INVALID_ARG_TYPE('fn', fn);
+            }
+            replacedImportMetaInitializer = true;
+            const parent = next;
+            next = (meta, context) => {
+              return fn(meta, context, parent);
+            };
+          });
+      } finally {
+        finished = true;
+        if (replacedImportMetaInitializer) {
+          this.#importMetaInitializer = next;
+        }
+      }
+    }
+  }
+
+  importMetaInitialize(meta, context) {
+    this.#importMetaInitializer(meta, context);
```

Comment:
Nit: is there any reason to keep the main part of this in a private field?

Reply:
That’s just how it is on current `main`:

https://github.com/nodejs/node/blob/4166d40d0873b6d8a0c7291872c8d20dc680b1d7/lib/internal/modules/esm/loader.js#L214

I’m open to refactoring, but maybe we could do so as part of the off-thread PR or after? I think that we’ll need to refactor `--experimental-import-meta-resolve` either as part of the off-thread PR or immediately thereafter. It should become sync like the browser version, using the `Atomics.wait` stuff in the off-thread PR to hook into our async resolve chain (or if there are no user loaders, it could map directly to `defaultResolve`).